File: Coding Problems & My Solutions/Maximum Subarray/maximum_subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int cur = nums[0];
        int max = cur;
        for (int i = 1; i < nums.length; i++){
            cur = Math.max(cur, 0);
            cur += nums[i];
            max = Math.max(max, cur);
        }
        return max;
    }
}

class Solution {
    public int maxSubArray(int[] nums) {
        int max = nums[0];
        int runningCount = max;
        for (int i = 1; i < nums.length; i++) {
            if (runningCount < 0) {
                runningCount = 0;
            }
            runningCount += nums[i];
            max = Math.max(max, runningCount);
        }
        return max;
    }
}

File: Coding Problems & My Solutions/Maximum Subarray/Maximum_Subarray.kt
class Solution {
    fun maxSubArray(nums: IntArray): Int {
        var maxVal = Integer.MIN_VALUE
        var count = 0
        for ((i, num) in nums.withIndex()) {
            if (count < 0) {
                count = 0
            }
            count += num
            maxVal = maxOf(maxVal, count)
        }
        return maxVal
    }
}

class Solution {
    fun maxSubArray(nums: IntArray): Int {
        var max = Int.MIN_VALUE
        if ( nums.size == 0 ){
            return max
        }
        else if ( nums.size == 1 ){
            return nums[0]
        }
        var i = 1
        var subMax = nums[0]
        max = Math.max(max, subMax)
        while ( i < nums.size ) {
            subMax = Math.max(nums[i], subMax + nums[i])      
            max = Math.max(max, subMax)
            i++
        }
        return max
    }
}